Nexus/Reconfigurator: Automatically adopt disks, deploy them to sleds via blueprints

smklein opened the pull request:
Automatically adopt disks and deploy them to sleds using blueprints

- Queries for physical disk info during reconfigurator planning phase
- Adds "physical disks" to blueprint, in-memory as well as the database schema
- Blueprint planning now ensures that in-service physical disks appear in the blueprint
- Blueprint execution sends a request to sled agents via `omicron-physical-disks PUT`
- A background task has been added to automatically adopt new physical disks as control plane objects, and to insert them into the database
- "Physical disk upsert" has largely been changed to "Physical disk insert", to avoid potential overwriting issues
- The physical disk "vendor/serial/model" uniqueness constraint has been removed for decommissioned disks. This will provide a pathway to eventually re-provisioning deleted disks, if an operator asks for it.

Still needs:
- [ ] Lots of tests.
  - [x] Fix the failing ones?
  - [x] Test background task
  - [ ] Test reconfigurator db changes
  - [x] Test reconfigurator planning
  - [x] Test disk deployment
  - [x] Test on real system?
- [x] Do I need to update how the "diff" tooling works? I updated the output of reconfigurator-cli, hopefully that's enough? 
- [x] This info should probably be plumbed into the "initial blueprint" via RSS (edit: it is now)

Fixes https://github.com/oxidecomputer/omicron/issues/5503 , https://github.com/oxidecomputer/omicron/issues/5502

jgallagher commented:
I realize there is urgency here and I don't want to create unnecessary work, but given this PR is already large-ish and there's a nontrivial list of things still to add (particularly tests of all the different pieces arriving here), would it make sense to break this into a few smaller PRs for easier reviewing / landing?  Just from reading the PR description it sounds like there might be some natural break points (planning input changes in one PR, database changes in one PR, new background task in one PR, maybe?), but if this isn't as natural as it sounds or is a bunch of work to organize feel free to veto this suggestion.

jgallagher commented on nexus/db-queries/src/db/datastore/deployment.rs:
```diff
@@ -412,6 +554,7 @@ impl DataStore {
             blueprint_zones,
+            blueprint_disks,
             parent_blueprint_id,
             internal_dns_version,
```
github won't let me put a comment on something that isn't in the diff, but `blueprint_delete` also needs to be updated to remove the disks associated with the blueprint-being-deleted.

jgallagher commented on nexus/db-queries/src/db/datastore/deployment.rs:
```diff
@@ -403,6 +495,56 @@ impl DataStore {
+                            // impossible and reflects either a bug or database
+                            // corruption.
+                            Error::internal_error(&format!(
+                                "disk {:?}: unknown sled: {:?}",
```
Small nit:

```suggestion
                                "disk {}: unknown sled: {}",
```

For `Uuid` this doesn't matter (its display and debug impls are [identical](https://docs.rs/uuid/1.8.0/src/uuid/fmt.rs.html#22-33)), but if/when we move these to `TypedUuid` it does matter (TypedUuid [includes the tag](https://docs.rs/newtype-uuid/1.0.1/src/newtype_uuid/lib.rs.html#155-158) in its debug impl, which I think we generally don't want in logs / error messages).

jgallagher commented on nexus/db-queries/src/db/datastore/deployment.rs:
```diff
@@ -403,6 +495,56 @@ impl DataStore {
+
+        // Sort all zones to match what blueprint builders do.
+        for (_, disks_config) in blueprint_disks.iter_mut() {
+            disks_config.disks.sort_by(|a, b| a.id.partial_cmp(&b.id).unwrap());
```
I think this would be equivalent and faster (allow unstable sorting, no unwrap needed):

```suggestion
            disks_config.disks.sort_unstable_by_key(|d| d.id);
```

jgallagher commented on nexus/db-queries/src/db/datastore/deployment.rs:
```diff
@@ -403,6 +495,56 @@ impl DataStore {
+            }
+        }
+
+        // Sort all zones to match what blueprint builders do.
```
```suggestion
        // Sort all disks to match what blueprint builders do.
```

jgallagher commented on nexus/db-queries/src/db/datastore/physical_disk.rs:
```diff
@@ -34,42 +38,30 @@ impl DataStore {
+            diesel::insert_into(dsl::physical_disk).values(disk.clone()),
         )
         .insert_and_get_result_async(conn)
         .await
```
Should the `ErrorHandler` in the .map_err after this line use `ErrorHandler::Conflict` now that this is an insert instead of an upsert?

jgallagher commented on nexus/db-queries/src/db/datastore/physical_disk.rs:
```diff
@@ -126,6 +118,82 @@ impl DataStore {
+        // but the sled policy crate explicitly avoids exporting that type.
+        //
+        // As a workaround, we must create the corresponding external types
+        // and manually convert them back into the database types.
```
I think this is intentionally difficult because we want to steer queries to using `SledFilter` is to avoid exactly this kind of "build up exactly the right policy / state combination" work. I think we can drop both of these, and change the filter to

```rust
.filter(SledFilter::InService)
```

jgallagher commented on nexus/db-queries/src/db/datastore/physical_disk.rs:
```diff
@@ -126,6 +118,82 @@ impl DataStore {
+                ),
+            )
+            // Filter out any disks in the inventory for which we have ever had
+            // a control plane disk.
```
This is probably a naive question because I don't grok the disk lifecycle yet, but: does this prevent us from reinitializing a disk that has previously been expunged?

andrewjstone replied:
Yes, I believe that is intentional based on last nights discussion, and possibly RFD 457.

jgallagher commented on nexus/src/app/background/physical_disk_adoption.rs:
```diff
@@ -0,0 +1,110 @@
+                        inv_disk.sled_id,
+                    );
+
+                    let result = self.datastore.physical_disk_insert(opctx, disk).await;
```
Just making sure I understand the flow; is this right?

1. A new sled is added
2. The sled-agent on the new sled creates zpools on all its physical disks
3. An inventory collection is taken when includes the new sled, which includes zpools+physical disks from step 2
4. This background task executes, which adopts any zpool+physicaldisk pairs, turning them into control plane `PhysicalDisks` via this call
5. The next time the planner runs, its `PlanningInput` includes the new sled with its zpool+physicaldisk pairs (it already would have gotten the new zpools; getting the physical disk alongside it is new)
6. The planner uses the new disk input to build `blueprint_disks`
7. The executor sends the list of zpool+disk pairs for each sled from `blueprint_disks` to every sled-agent (an idempotent operation)

This gets at the question of autoadoption vs operator-driven adoption - it feels a little backwards to me that the disks are adopted _before_ reconfigurator gets involved, and then execution is only responsible for telling sled-agents about it. I think the feeling is because for external networking resources, reconfigurator chooses whether to assign resources during planning, and then execution is responsible for both (a) record that decision in the database and then (b) tell sled-agent about it, whereas for disks (a) is happening ahead of time. Maybe that's fine or even better in this case? I'm not sure, but just wanted to jot down the feeling.

andrewjstone replied:
I think it makes sense to auto adopt in this case. We need a mechanism to create the physical control plane disks so that reconfigurator can see them. This is somewhat of a stop-gap because we don't have APIs to list raw disks and allow the operator to adopt them. However, as an intermediate step, we could choose to have the physical control plane disks not put in service and then adopted by reconfigurator. I wouldn't worry about that for this PR though.

andrewjstone commented on nexus/reconfigurator/execution/src/omicron_physical_disks.rs:
```diff
@@ -0,0 +1,265 @@
+    sleds_by_id: &BTreeMap<Uuid, Sled>,
+    sled_configs: &BTreeMap<TypedUuid<SledKind>, BlueprintPhysicalDisksConfig>,
+) -> Result<(), Vec<anyhow::Error>> {
+    let errors: Vec<_> = stream::iter(sled_configs)
```
I think this is purely sequential. Do we want to instead make deployment concurrent via `FuturesUnordered` or the like? My main concern is multiple timeouts adding up consecutively, rather than amortized to a single timeout.

jgallagher replied:
I think I'd vote for keeping these sequential on this PR, and have a separate issue for blueprint realization concurrency / optimization. I'd maybe also group "error handling" in that (right now, there are several ways an error can prevent further blueprint realization; I think that's probably overly cautious, but relaxing it might be tricky).

andrewjstone commented on nexus/reconfigurator/planning/src/blueprint_builder.rs:
```diff
@@ -405,6 +438,89 @@ impl<'a> BlueprintBuilder<'a> {
         self.comments.push(String::from(comment));
     }
 
+    /// Ensures that the blueprint contains disks for a sled which already
```
This method is super clear and well documented. Thanks!

andrewjstone commented on nexus/reconfigurator/planning/src/blueprint_builder.rs:
```diff
@@ -405,6 +438,89 @@ impl<'a> BlueprintBuilder<'a> {
+            )
+        });
+
+        Ok(Ensure::Added)
```
Nit: maybe not worth changing right now since this is used elsewhere, but `Ensure::Modified` is perhaps a better name.

andrewjstone commented on nexus/reconfigurator/planning/src/blueprint_builder.rs:
```diff
@@ -863,6 +981,102 @@ impl<'a> BlueprintZonesBuilder<'a> {
+///
+/// Tracking the set of disks is slightly non-trivial because we need to bump
+/// the per-sled generation number iff the disks are changed.  So we need to
+/// keep track of whether we've changed the disksrelative to the parent
```
```suggestion
/// keep track of whether we've changed the disks relative to the parent
```

andrewjstone reviewed (commented):
Thanks for jumping on this so quickly @smklein. It's a great start!

smklein commented:
Also - I'm on board to split this up! But I might do it as a late-stage step taking things out of draft, because it's easier to tell when I'm "green on CI" with a single PR, even if logically I can split things into smaller chunks.

smklein commented:
Tested this on a4x2, appears to be working! @andrewjstone is following-up on an issue to get inventory to re-trigger more quickly when sled add completes (It *is* triggering right now, but before the sled agent is in the DB, so we don't actually query the new sled for inventory info yet)

andrewjstone reviewed (approved):
Ship It! Tested on a4x2 and huzzah!

andrewjstone commented:
> Tested this on a4x2, appears to be working! @andrewjstone is following-up on an issue to get inventory to re-trigger more quickly when sled add completes (It _is_ triggering right now, but before the sled agent is in the DB, so we don't actually query the new sled for inventory info yet)

Work for this is in https://github.com/oxidecomputer/omicron/pull/5526

smklein commented:
One last minute change, FYI -- I changed zpool / physical disk "upsert" to "insert" a while ago, because they are now creating control plane objects, rather than "updating a view of info from sleds". That's the responsibility of the inventory now.

Using INSERT here is the right call, **most** of the time, except for one spot: RSS, which wants to "INSERT unless it already exists". Normally, INSERT throws an error if there's a conflict. RSS would want to explicitly keep going. I fixed that here: https://github.com/oxidecomputer/omicron/pull/5506/commits/c985ddcb72f3c1bdb0aec029605770e5db1d207d when I saw another a4x2 run get stuck after trying to re-insert these devices.
